src/input.rs

```diff
@@ -0,0 +1,232 @@
+    pub fn new() -> Self {
+        Self { x: 0.0, y: 0.0 }
+    }
+}
+
+/// Information about KeyboardKey or MouseButton
+#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone)] //, Serialize, Deserialize)]
+pub enum Button {
+    Key(VirtualKeyCode), // TODO: add support for Key{scancode: u32}?
+    Mouse(MouseButton),
+    None,
+}
+
+/// Input Service
+pub struct Input {
+    mapper: Box<dyn std::any::Any + Send + Sync>,
+    states: HashMap<Button, State>,
+    mouse_scroll_delta: f32,
+    mouse_position: Vector2,
+}
+
+impl Input {
+    pub fn new(mapper: Box<dyn std::any::Any + Send + Sync>) -> Self {
+        Self {
+            mapper,
+            states: HashMap::new(),
+            mouse_scroll_delta: 0.0,
+            mouse_position: Vector2::new(),
+        }
+    }
+
+    /// Checks if user defined action button is pressed
+    pub fn action_pressed<T>(&self, action: T) -> bool // TODO: reduce code (almost same as action_hold)
+    where
+        Self: ActionMapper<T>,
+        T: Eq + std::hash::Hash,
+    {
+        if let Some(button) = self.get_action(action) {
+            if let Some(state) = self.states.get(button) {
+                return *state == State::Pressed;
+            }
+        }
+        false
+    }
+
+    /// Checks if user defined action button is released
+    pub fn action_released<T>(&self, action: T) -> bool // TODO: reduce code (almost same as action_hold)
+    where
+        Self: ActionMapper<T>,
+        T: Eq + std::hash::Hash,
+    {
+        if let Some(button) = self.get_action(action) {
+            if let Some(state) = self.states.get(button) {
+                return *state == State::Released;
+            }
+        }
+        false
+    }
+
+    /// Checks if user defined action button is pressed or hold
+    pub fn action_hold<T>(&self, action: T) -> bool
+    where
+        Self: ActionMapper<T>,
+        T: Eq + std::hash::Hash,
+    {
+        if let Some(button) = self.get_action(action) {
+            if let Some(state) = self.states.get(button) {
+                return *state == State::Hold || *state == State::Pressed;
+            }
+        }
+        false
+    }
+
+    /// just a mapper getter
+    pub fn mapper<T: 'static + Send + Sync>(&self) -> &T {
+        self.mapper.downcast_ref::<T>().unwrap()
+    }
+
+    /// This method should be called in application.rs before handle_event, so States from events will be properly updated
+    pub fn handle_update(&mut self) {
+        self.mouse_scroll_delta = 0.0;
+
+        self.states.retain(|_btn, state| match state {
+            State::Pressed => {
+                *state = State::Hold;
+                true
+            }
+            State::Released => false,
+            _ => true,
+        });
+    }
+
+    /// This method should be called in application.rs before the main match case, so systems will get actual Input data
+    pub fn handle_event(&mut self, event: &winit::event::Event<'_, ()>) {
```

Comment:
There is one global issue with it: `handle_event` could be called several times between renderings (and standard systems triggering), it means, that Input manager should aggregate inputs and reset them once rendering is done.